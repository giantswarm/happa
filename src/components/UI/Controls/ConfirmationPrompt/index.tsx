import { Box, Collapsible, Keyboard, Text } from 'grommet';
import PropTypes from 'prop-types';
import React, { useEffect, useRef } from 'react';

import Button from '../Button';

interface IConfirmationPromptProps
  extends Omit<React.ComponentPropsWithRef<typeof Collapsible>, 'title'> {
  title?: React.ReactNode;
  confirmButton?: React.ReactNode;
  cancelButton?: React.ReactNode;
  onConfirm?: () => void;
  onCancel?: () => void;
}

const ConfirmationPrompt = React.forwardRef<
  HTMLDivElement,
  IConfirmationPromptProps
>(
  (
    {
      children,
      title,
      confirmButton,
      cancelButton,
      onConfirm,
      onCancel,
      ...props
    },
    ref
  ) => {
    const wrapperRef = useRef<HTMLDivElement>(null);
    const cancelButtonRef = useRef<HTMLButtonElement>(null);

    useEffect(() => {
      /**
       * Focus a control after the confirmation is opened and
       * all the microtasks have finished running.
       *  */
      setTimeout(() => {
        if (!props.open) return;

        if (cancelButtonRef.current) {
          // Focus on the cancel button if it exists.
          const cancelButtonElement = cancelButtonRef.current.querySelector<HTMLButtonElement>(
            '.cancel-button'
          );
          cancelButtonElement?.focus();
        } else if (wrapperRef.current) {
          // Otherwise focus on the wrapper element.
          wrapperRef.current.focus();
        }
      });
    }, [props.open]);

    const handleConfirm = (e: React.MouseEvent<HTMLElement>) => {
      e.preventDefault();

      onConfirm?.();
    };

    const handleCancel = (
      e: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLElement>
    ) => {
      e.preventDefault();

      onCancel?.();
    };

    return (
      <Collapsible {...props} ref={ref}>
        <Keyboard onEsc={handleCancel}>
          <Box
            ref={wrapperRef}
            background='background-front'
            round='xsmall'
            pad='medium'
            tabIndex={-1}
          >
            {typeof title === 'string' ? (
              <Text weight='bold' margin={{ bottom: 'small' }}>
                {title}
              </Text>
            ) : (
              title
            )}

            {children}

            {(onConfirm || onCancel) && (
              <Box
                direction='row'
                margin={{ top: 'medium' }}
                justify='center'
                gap='small'
              >
                {onConfirm &&
                  (typeof confirmButton === 'string' ? (
                    <Button danger={true} onClick={handleConfirm}>
                      {confirmButton}
                    </Button>
                  ) : (
                    confirmButton
                  ))}

                {onCancel &&
                  (typeof cancelButton === 'string' ? (
                    <Button
                      onClick={handleCancel}
                      ref={cancelButtonRef}
                      className='cancel-button'
                    >
                      {cancelButton}
                    </Button>
                  ) : (
                    cancelButton
                  ))}
              </Box>
            )}
          </Box>
        </Keyboard>
      </Collapsible>
    );
  }
);

ConfirmationPrompt.propTypes = {
  open: PropTypes.bool,
  children: PropTypes.node,
  title: PropTypes.node,
  confirmButton: PropTypes.node,
  cancelButton: PropTypes.node,
  onConfirm: PropTypes.func,
  onCancel: PropTypes.func,
};

ConfirmationPrompt.defaultProps = {
  confirmButton: 'Confirm',
  cancelButton: 'Cancel',
};

export default ConfirmationPrompt;
