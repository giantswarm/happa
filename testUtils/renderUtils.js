import { render } from '@testing-library/react';
import App from 'App';
import { ConnectedRouter } from 'connected-react-router';
import { ThemeProvider } from 'emotion-theming';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { AppRoutes } from 'shared/constants/routes';
import configureStore from 'stores/configureStore';
import theme from 'styles/theme';

export const initialStorage = {
  user:
    '{"email":"developer@giantswarm.io","auth":{"scheme":"giantswarm","token":"a-valid-token"},"isAdmin":true}',
};

/**
 * This function will render the whole app with a mocked store in the route
 * provided.
 * @param {String} route.
 * @param {HTMLElement} container An HTMLElement.
 */
export function renderRouteWithStore(
  initialRoute = AppRoutes.Home,
  state = {},
  storage = initialStorage
) {
  localStorage.replaceWith(storage);

  const history = createMemoryHistory({
    initialEntries: [initialRoute],
    initialIndex: 0,
  });

  const store = configureStore(state, history);

  const app = render(<App {...{ store, theme, history }} />);

  return app;
}

/**
 * Render component in the Theme context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithTheme(component, props, options) {
  return render(getComponentWithTheme(component, props), options);
}

/**
 * Include component in the Theme context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 */
export function getComponentWithTheme(Component, props) {
  return (
    <ThemeProvider theme={theme}>
      <Component {...props} />
    </ThemeProvider>
  );
}

/**
 * Render component in the Store context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithStore(component, props, state, options) {
  return render(getComponentWithStore(component, props, state), options);
}

/**
 * Include component in the Store context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {Record<string, any>} storage Current LocalStorage value
 * @param {History} history Current Browser history
 */
export function getComponentWithStore(
  Component,
  props = {},
  state = {},
  storage = initialStorage,
  history = createMemoryHistory()
) {
  localStorage.replaceWith(storage);

  const store = configureStore(state, history);

  const app = (
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <ConnectedRouter history={history}>
          <Component {...props} />
        </ConnectedRouter>
      </ThemeProvider>
    </Provider>
  );

  return app;
}
