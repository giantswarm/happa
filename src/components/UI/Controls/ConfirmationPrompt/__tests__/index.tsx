import { renderWithTheme } from 'test/renderUtils';

import ConfirmationPrompt from '..';

describe('ConfirmationPrompt', () => {
  it('renders a simple component', () => {
    const { container } = renderWithTheme(ConfirmationPrompt, {
      children: 'Hi friends',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a component with buttons', () => {
    const { container } = renderWithTheme(ConfirmationPrompt, {
      children: 'Hi friends',
      onConfirm: () => {},
      onCancel: () => {},
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a component with a title', () => {
    const { container } = renderWithTheme(ConfirmationPrompt, {
      children: 'Welcome to my prompt',
      title: 'Hi friends',
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
