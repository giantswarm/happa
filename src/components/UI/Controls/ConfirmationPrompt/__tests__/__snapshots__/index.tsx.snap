// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmationPrompt renders a component with a title 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 giQQcM"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 kuNpSX"
  >
    <div
      aria-hidden="true"
      class="StyledBox-sc-13pk1d4-0 hpSXsQ Collapsible__AnimatedBox-sc-15kniua-0 jtbjrr"
    />
  </div>
</div>
`;

exports[`ConfirmationPrompt renders a component with buttons 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 giQQcM"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 kuNpSX"
  >
    <div
      aria-hidden="true"
      class="StyledBox-sc-13pk1d4-0 hpSXsQ Collapsible__AnimatedBox-sc-15kniua-0 jtbjrr"
    />
  </div>
</div>
`;

exports[`ConfirmationPrompt renders a simple component 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 giQQcM"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 kuNpSX"
  >
    <div
      aria-hidden="true"
      class="StyledBox-sc-13pk1d4-0 hpSXsQ Collapsible__AnimatedBox-sc-15kniua-0 jtbjrr"
    />
  </div>
</div>
`;
